code/modules/recycling/disposal-construction.dm

```diff
@@ -170,10 +170,13 @@
 			nicetype = "pipe"
 
 	var/turf/T = loc
-	if(T.intact)
+	if(T.intact && istype(T, /turf/simulated/floor))
 		user << "You can only attach the [nicetype] if the floor plating is removed."
 		return
-
+	if(!ispipe)
+		user << "You can't build [nicetype]s on walls, only disposal pipes."
+		return

```

Comment:
I thought I tested building bins and stuff after this change, I'll look at this later when I get home.